rest_specification/restful_service_specification.md

```diff
@@ -0,0 +1,240 @@
+| `DELETE`       | Delete the resource           | -                | -            | -             |
+
+Clients should use the `fields` parameter for better performance.
+
+Implementors should consider breaking out field represented by a JSON array
+into sub-entities if there are performance concerns around its size. This
+consideration should happen regardless of how the data is stored.
+
+## Pagination
+
+Paging shall be implemented via `page` and `size` query parameters. If none are
+provided, the default page size shall be used.
+
+The default page-size is implementation-specific.
+
+| Parameter | Description               |
+|-----------|---------------------------|
+| `page`    | page index, starting at 0 |
+| `size`    | page size                 |
+
+## Sorting
+
+Sorting shall be provided via the `sort` query parameter. `sort` may be
+specified more than once in order to sort by more than one field. `sort` values
+shall be of the form `sort={field}(,{direction})` where `direction` is `asc` or
+`desc`.
+
+### Responses
+
+The response shall be a JSON object formatted according to Spring Data REST
+conventions.Specifically, the list is returned as a pluralized field name
+under `_embedded`, and a `page` object supplies paging info. Example:
+
+```json
+GET /user
+
+{
+  "_embedded": {
+    "users": [
+      {
+        ...
+      },
+      ...
+    ]
+  },
+  "page": {
+    "size": 20,
+    "totalElements": 200,
+    "totalPages": 10,
+    "number": 0
+  }
+}
+```
+
+## Additional Functions
+
+Additional APIs may expose other functionality related to a resource type.
+Examples include:
+ - query for a specific subset of resources
+ - perform an action against a set of resources
+
+Additional functions shall be `POST` only, and parameters will be provided in
+the request body. The request body SHOULD be a JSON object with fields
+corresponding to parameter names.
+Examples:
+
+```json
+POST /user/findByLogin
+
+{
+  "login": "bob"
+}
+```
+
+Each function that can be performed against a specific resource must exist in
+two forms:
+ - non-batch
+ - batch
+
+The non-batch form shall take any relevant parameters in a JSON object. Example:
+
+```json
+POST /user/123/deactivate
+
+{
+  "reason": "Banned"
+}
+```
+
+The batch form shall take a JSON array of objects, with each object having the
+identifier as an additional field. Example:
+
+```json
+POST /user/deactivate
+
+[
+  {
+    "id": 123,
+    "reason": "Banned"
+  },
+  {
+    "id": 456,
+    "reason": "Banned"
+  }
+]
+```
+
+Responses for functions shall follow [JSend](http://labs.omniti.com/labs/jsend)
+conventions:
+ - Responses must be JSON objects
+ - `status` field shall tell whether the function call was successful
+  - possible values: `success`, `fail`, `error`
+ - `data` shall have any feedback from the function call (function-specific)
+ - `message` shall contain a descriptive error message in the case of an error.
```

Comment:
Sometimes there may be multiple errors. It might be nice if this could be an array. It might also be nice if this could be used for warnings also (such as when someone is using a deprecated code path for some reason or another). I've seen some API do this (I think it was Rally's) and found it especially useful when exploring the API to figure out what requests I needed to make.